src/Step.cc

```diff
@@ -122,38 +122,34 @@ bool Step::execute(Context& context, CommChannel* comm, StepIndex index)
 
     copy_used_variables_from_context_to_lua(context, lua);
 
-    bool result = false;
+    const auto result_or_error = execute_lua_script_safely(lua, get_script());
 
-    try {
-        auto protected_result = lua.safe_script(get_script(), sol::script_default_on_error);
-        if (!protected_result.valid()) {
-            sol::error err = protected_result;
-            throw ErrorAtIndex(cat("Script execution error: ", err.what()), index);
-        }
-
-        copy_used_variables_from_lua_to_context(lua, context);
-        sol::optional<bool> opt_result = protected_result;
-        if (opt_result)
-            result = *opt_result;
-    }
-    catch (const std::exception& e) {
-        std::string msg = cat("Script execution error: ", e.what());
-        send_message(comm, Message::Type::step_stopped_with_error, msg,
-            Clock::now(), index);
-        throw ErrorAtIndex(msg, index);
-    }
-    catch (...) {
-        std::string msg{ "Script execution raised unknown exception" };
-        send_message(comm, Message::Type::step_stopped_with_error, msg,
-            Clock::now(), index);
+    if (std::holds_alternative<std::string>(result_or_error))
+    {
+        const auto msg = std::get<std::string>(result_or_error);
+        send_message(comm, Message::Type::step_stopped_with_error, msg, Clock::now(),
+                     index);
         throw ErrorAtIndex(msg, index);
     }
 
+    const auto result_obj = std::get<sol::object>(result_or_error);
+
+    const bool result_bool = [](const sol::object& obj) -> bool
+        {
+            if (obj == sol::nil)          return false;
```

Comment:
That's right, I was going for the C way and forgot about the Lua way. We should discuss what we want; we could also enforce a boolean return value if one is needed for the step type and throw an error otherwise.

Reply:
> we could also enforce a boolean return value

Maybe that is the best solution. Enforce script writers to return a bool on IF and WHILE step scripts.
That would make sure that they do neither assume C logic nor Lua logic, which is incompatible for zeros.

As we can not do static type checking, we must instead throw an error if an unexpected type is returned.

Reply:
After some more thinking

* Why is a change in `execute()` behavior (what es reported as `true` or `false`) changed in THIS PR at all? I would separate that out into a new PR, apart from exceptions.
* Following natural flow a `Step` does not know about a `StepIndex`. Why do we leak the concept of `Sequence` into a simple `Step`? This is completely unneeded.

About the 2nd point:
* `Sequences` handle `Steps`, `Sequences` must know something about `Steps`. `Steps` on the other hand could be used on their own (without a `Sequence`). For good encapsulation `Steps` should not care about `Sequences`.
* But we have a coupling ... you can not execute a `Step` without giving it at least a fake `index`. Hmm.
* The `Step` uses the index solely to produce `Messages`.
* Why does it need to produce a message? That is a design choice.
* If you run simple single `Steps` (without a `Sequence`) you do not need the `Message` mechanism at all.
* Here is another take on that design and where to put the border:
  * A `Step` is dump. A Step does not throw. A Step does not message.
  * In fact a `Step` does not even have a `Step::Type`, it is not evaluated there at all. It's just convenient for us that it holds that as metadata (as is the case with indent etc).
  * `Step::execute()` just returns the `std::variant<sol::object, std::string>` (which I would abstract away as for ex `LuaExecResult` alias.
  * The `Sequence` knows which `Step` it is about to start, which `Step` just crashed or returned. It knows the index.
  * The `Sequence` can create the `Messages` before starting and after returning from a `Step` execution.
  * The `Sequence` can decide if the returned value is `ok`; which depends on the `Step::Type`.

So ... my proposal is:
* Rip messaging and return value evaluation out of `Step` and put it into `Sequence`
* `Sequence` can enforce `bool` return values for `Step::Type::IF` and `Step::Type::WHILE`, while keeping all other types free to return nothing or anything.

I came to this conclusion while I changed the code to implement 'require Steps to return a bool on IF and WHILE while other types can return what they want'.
Of course one can implement that purely in `Step.cc`, but I have the feeling that the abstraction is broken. How this develops the difference between `Sequence` and `Steps` and what they have to care about gets weaker and weaker. In order to behave correctly the `Step` must know stuff about `Sequences`. Which types should be handled how in a sequence. For me it is a difference if the `Step` carries some data for the `Sequence` and does not know what the data means, and the `Step` takes the `Sequences` information and incorporates it into its own behavior.

Reply:
> Why is a change in execute() behavior (what es reported as true or false) changed in THIS PR at all? I would separate that out into a new PR, apart from exceptions.

That would have been better, I agree. But the behavior change came as a side-effect of the refactoring around `execute_lua_script_safely()`. I have now addressed that in an additional commit that establishes the desired behavior.

As for better logical separation of Step and Sequence, I'd have to think about that next week.

Reply:
About abstraction and layers: What was on my mind while writing this has been a discussion with a programming Flash-D colleague who just 'discovered' _Clean Code_ and I was like \*cringe\* and recommended as a counterexample _A Philosophy of Software Design_ [(review)](https://blog.pragmaticengineer.com/a-philosophy-of-software-design-review/).

Because we discussed that a lot in the tea-kitchen (like: recently) the arguments of Outerhout are still present although I bought that book quite a while ago. Now he borrowed it and I can not even cite :grimacing: 

Maybe I should present something on that topic in the lectures series ;-D

* `git` (in theory)
* `git` possibilities on the command line
* How to github
* How to gitlab (ugg, not me)
* On software design (send some impulses) not as lecture but present some ideas

Reply:
> Here is another take on that design and where to put the border:
>    A Step is dump. A Step does not throw. A Step does not message.
 >   In fact a Step does not even have a Step::Type, it is not evaluated there at all. It's just convenient for us that it holds that as metadata (as is the case with indent etc).
>    Step::execute() just returns the std::variant<sol::object, std::string> (which I would abstract away as for ex LuaExecResult alias.
>    The Sequence knows which Step it is about to start, which Step just crashed or returned. It knows the index.
>    The Sequence can create the Messages before starting and after returning from a Step execution.
>    The Sequence can decide if the returned value is ok; which depends on the Step::Type.

I like the idea. Injecting the index really is a "Notnagel" and bugs me every time I see it. I am not entirely sure about messaging, though: Also the `print` command needs to send messages, after all. Sure, it gets injected via the `Context`, but we need to see how to give it a reference to the `CommChannel`. I'll noodle on it a bit...

Reply:
About the lecture series: I am sure Tim would love to restart that at some point. I can throw in one or two talks as well.

Reply:
Silly idea:
* The `Step` send a message to the `CommChannel`
* The `Sequence` hooked into the `CommChannel`
* The `Sequence` callback sees the `Step`'s Message and appends `index` information
* The `Executor` (or in direct mode: the user themselves) receives finally the message